SB-215 (feat) : 중고 시스템 전체 리팩토링

seob7 opened the pull request:
# SB-215 (feat) : 중고 시스템 전체 리팩토링

## 변경 사항

### 기존 시스템

- **중고 물품 등록** : 중고 물품 등록을 수행할 때 `GoodsStatus`만 변경 (STORAGE -> USED)
- **중고 물품 폼 작성** : 중고 물품에 대한 상세 정보를 작성하는 프로세스

### 변경된 시스템

- **중고 물품 등록 통합**: 중고 물품 폼 작성 시 `GoodsStatus`도 동시에 변경되도록 개선하였습니다.

## API

| **변경 전**                                          | **변경 후**                                        |
|-----------------------------------------------------|---------------------------------------------------|
| **중고 물품 등록**                                  | **중고 물품 등록**                                |
| `POST : "/api/usedgoods/{goodsId}"`                  | `POST : "/api/usedgoods"`                         |
| **중고 물품 등록 취소**                             | **중고 물품 등록 취소**                           |
| `POST : "/api/usedgoods/cancel/{goodsId}"`           | `POST : "/api/usedgoods/cancel/{usedGoodsId}"`    |
| **중고 물품 폼 작성**                               | **중고 물품 등록과 통합**                         |
| `POST : "/api/usedgoods/post"`                       | -                                                 |

## DB

- **status 필드 추가** : `REGISTERED`, `UNREGISTERED`, `SOLD`
- **goodsId 필드 추가**

## 클래스

| **구분**      | **변경 전**                | **변경 후**                  |
|---------------|----------------------------|------------------------------|
| **Request**   | `UsedGoodsPostRequest`     | `RegisterUsedGoodsRequest`   |
|               |                            | `CancelUsedGoodsRequest` 추가 |
| **Response**  | `TransformUsedGoodsResponse`|  제거       |
|               |                `UsedGoodsPostResponse`            | `UsedGoodsDetailResponse`  |
|               |                            | `MessageResponse` 추가 |
| **Status**    |                            | `UsedGoodsStatus` 추가       |
| **Exception** |                 | `UsedGoodsNotFoundException` 추가 |
|               |                            | `UsedGoodsExceptionHandler` 추가 |
|              |                             |                         `UsedGoodsErrorCode` 추가                    | 



## 추가된 기능

### `usedGoodsId`로 중고 물품 구매

- **URL**: `/api/usedgoods/{usedGoodsId}`
- **Method**: `POST`

**프로세스는 다음과 같습니다. (추후 변경 예정)**

1. `usedGoodsStatus`가 `REGISTERED`인지 확인
2. `goods`의 `userId`를 구매자의 `userId`로 변경
3. `goods`의 `status`를 보관(STORAGE)로 변경
4. `usedGoodsStatus`를 `SOLD`로 변경

```
// TODO 구매 프로세스 재정립
public MessageResponse buyUsedGoods(Long usedGoodsId, String email) {

    Long userId = usersService.getUserWithThrow(email).getId();

    UsedGoodsEntity usedGoodsEntity = usedGoodsService.getUsedGoodsBy(usedGoodsId,
        UsedGoodsStatus.REGISTERED); // 1

    GoodsEntity goodsEntity = goodsService.getGoodsById(usedGoodsEntity.getGoodsId());

    goodsService.setUserId(goodsEntity, userId); // 2

    goodsService.setGoodsStatusBy(goodsEntity, GoodsStatus.STORAGE); // 3

    usedGoodsService.setUsedGoodsStatusBy(usedGoodsEntity, UsedGoodsStatus.SOLD); // 4

    return usedGoodsConverter.toMessageResponse("중고 물품 거래가 완료되었습니다.");
}

seob7 committed:
SB-215 (feat) : 중고 시스템 전체 리팩토링

SB-215 (feat) : 중고 시스템 전체 리팩토링

변경 사항
기존 시스템
- 중고 물품 등록 : 중고 물품 등록을 수행할 때 GoodsStatus 만 변경 (STORAGE -> USED)
- 중고 물품 폼 작성 : 중고 물품에 대한 상세 정보를 작성하는 프로세스

변경된 시스템
- 중고 물품 등록 통합 : 중고 물품 폼 작성 시 GoodsStatus도 동시에 변경되도록 개선하였습니다.

추가된 기능
- usedGoodsId로 중고 물품 구매

shinywoon commented on db/src/main/java/db/domain/usedgoods/enums/UsedGoodsStatus.java:
```diff
@@ -0,0 +1,19 @@
+
+    REGISTERED(1,"중고 물품 등록", "판매 중인 중고 물품입니다."),
+    UNREGISTERED(2,"중고 물품 취소", "판매 취소된 중고 물품입니다."),
+    SOLD(3, "중고 거래 완료", "중고 물품 거래가 완료되었습니다.")
```
저희 Process 에 의하면 더 많은 상태가 필요합니다.
REGISTERED , UNREGISTERED , DEALING, DEAL, REMITTANCE, ASSIGNMENT ,SOLD
따라서 기능은
```
1. 중고 전환하기
2. 중고 전환 취소하기
3. 중고 거래하기
4. 중고 거래 확정하기
5. 송금 확인
4. 물건 확인
5. 완료
```
위 Process 를 따라야 합니다.

shinywoon commented on warehouse/src/main/java/warehouse/common/error/UsedGoodsErrorCode.java:
```diff
@@ -0,0 +1,19 @@
+@AllArgsConstructor
+public enum UsedGoodsErrorCode implements ErrorCodeIfs {
+
+    USED_GOODS_NOT_FOUND(HttpStatus.NOT_FOUND.value(), 1600, "중고 물품을 찾을 수 없습니다.")
```
1. 물품 자체가 존재하지 않는 경우
2. 물품 상태가 중고 전환 상태가 아닌경우
위 2가지 경우 모두 고려해주세요!

shinywoon reviewed (changes requested):
코드는 너무 좋습니다.
다만 리뷰 확인 하시고 수정 요청드립니다.

seob7 committed:
SB-215 (refactor) : UsedGoodsStatus, ErrorCode 추가

SB-215 (refactor) : UsedGoodsStatus, ErrorCode 추가

1. UsedGoodsStatus 에 DEALING, DEAL, REMITTANCE, ASSIGNMENT, SOLD 프로세스 추가
2. 물품이 중고 전환 상태인지 확인하는 ErrorCode 및 ExceptionHandler 구현
3. UsedGoodsService
    - UsedGoodsNotFound -> UsedNotInUsedStatus 로 예외코드 변경

seob7 committed:
Merge branch 'dev' into refactor/usedgoods

seob7 committed:
Merge branch 'dev' into refactor/usedgoods
